java-checks/src/main/java/org/sonar/java/checks/OptionalAsParameterCheck.java

```diff
@@ -0,0 +1,78 @@
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+
+import org.sonar.check.Rule;
+import org.sonar.java.resolve.JavaType;
+import org.sonar.java.resolve.ParametrizedTypeJavaType;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+
+import java.util.List;
+import java.util.Optional;
+
+@Rule(key = "S3553")
+public class OptionalAsParameterCheck extends IssuableSubscriptionVisitor {
+
+  private static final String JAVA_UTIL_OPTIONAL = "java.util.Optional";
+  private static final List<String> PRIMITIVE_OPTIONALS = ImmutableList.<String>builder()
+    .add("java.util.OptionalDouble")
+    .add("java.util.OptionalInt")
+    .add("java.util.OptionalLong")
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    for (VariableTree parameter : ((MethodTree) tree).parameters()) {
+      TypeTree typeTree = parameter.type();
+      String expectedType = expectedTypeInsteadOfOptional(typeTree.symbolType());
+      if (expectedType != null) {
+        reportIssue(typeTree, "Specify a \"" + expectedType + "\" parameter instead.");
+      }
+    }
+  }
+
+  @CheckForNull
+  private String expectedTypeInsteadOfOptional(Type type) {
+    if (type.is(JAVA_UTIL_OPTIONAL) && ((JavaType) type).isParameterized()) {
+      ParametrizedTypeJavaType parametrizedTypeJavaType = (ParametrizedTypeJavaType) type;
+      return parametrizedTypeJavaType.substitution(parametrizedTypeJavaType.typeParameters().get(0)).erasure().name();
+    }
+    Optional<String> primitiveOptional = PRIMITIVE_OPTIONALS.stream().filter(optional -> type.is(optional)).findFirst();
+    if (primitiveOptional.isPresent()) {
+      return primitiveOptional.get().substring(JAVA_UTIL_OPTIONAL.length()).toLowerCase();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define the locale to be used in this String operation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1449)